File: app/src/main/java/edu/uofk/eeese/eeese/projects/ProjectsPresenter.java
/*
 * Copyright 2017 Ali Salah Alddin
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package edu.uofk.eeese.eeese.projects;

import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import edu.uofk.eeese.eeese.data.Project;
import edu.uofk.eeese.eeese.data.source.BaseDataRepository;
import edu.uofk.eeese.eeese.util.schedulers.BaseSchedulerProvider;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class ProjectsPresenter implements ProjectsContract.Presenter {

    @NonNull
    private BaseDataRepository mSource;
    @NonNull
    private ProjectsContract.View mView;
    @NonNull
    private BaseSchedulerProvider mSchedulerProvider;
    private
    @Project.ProjectCategory
    int mCategory;

    private CompositeDisposable mSubscriptions;

    @Inject
    public ProjectsPresenter(@NonNull BaseDataRepository source,
                             @NonNull ProjectsContract.View view,
                             @NonNull BaseSchedulerProvider schedulerProvider,
                             @Project.ProjectCategory int category) {
        mSource = source;
        mView = view;
        mSchedulerProvider = schedulerProvider;
        mCategory = category;
        mSubscriptions = new CompositeDisposable();
    }

    @Override
    public int getCategory() {
        return mCategory;
    }

    @Override
    public void loadProjects(final boolean force) {
        mView.setLoadingIndicator(true);
        Disposable subscription =
                mSource.getProjectsWithCategory(force, mCategory)
                        .subscribeOn(mSchedulerProvider.io())
                        // sort by name
                        .map(projects -> {
                            List<Project> sortedProjects = new ArrayList<>(projects);
                            Collections.sort(
                                    sortedProjects,
                                    (lhs, rhs) -> lhs.getName().compareToIgnoreCase(rhs.getName())
                            );
                            return sortedProjects;
                        })
                        .subscribeOn(mSchedulerProvider.computation())
                        .observeOn(mSchedulerProvider.ui())
                        .doFinally(() -> mView.setLoadingIndicator(false))
                        .subscribe(
                                // OnSuccess
                                projects -> {
                                    if (projects.size() > 0) {
                                        mView.showProjects(projects);
                                    } else {
                                        mView.showNoProjects();
                                    }
                                },
                                // OnError
                                throwable -> {
                                    mView.showNoConnectionError();
                                    mView.showNoProjects();
                                }
                        );
        mSubscriptions.add(subscription);
    }

    @Override
    public void openProjectDetails(@NonNull Project project) {
        mView.showProjectDetails(project.getId());
    }

    @Override
    public void subscribe() {
    }

    @Override
    public void unsubscribe() {
        mSubscriptions.clear();
    }
}

File: app/src/main/java/edu/uofk/eeese/eeese/projects/ProjectsPresenter.kt
/*
 * Copyright 2017 Ali Salah Alddin
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package edu.uofk.eeese.eeese.projects

import edu.uofk.eeese.eeese.data.Project
import edu.uofk.eeese.eeese.data.ProjectCategory
import edu.uofk.eeese.eeese.data.source.ProjectWithCategory
import edu.uofk.eeese.eeese.data.source.Repository
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject


class ProjectsPresenter @Inject
constructor(private val source: Repository<Project>,
            private val view: ProjectsContract.View,
            private val category: ProjectCategory) : ProjectsContract.Presenter {

    private val subscriptions: CompositeDisposable = CompositeDisposable()

    override fun getCategory(): ProjectCategory {
        return category
    }

    override fun loadProjects(force: Boolean) {
        if (force) {
            view.setLoadingIndicator(true)
            val sync = source.sync().subscribe()
            subscriptions.add(sync)
        }
        val subscription = source.get(ProjectWithCategory(category))
                .subscribeOn(Schedulers.io())
                // sort by name
                .map({ projects ->
                    projects.toList().sortedBy { it.name }
                })
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnEach { view.setLoadingIndicator(false) }
                .subscribe(
                        // OnSuccess
                        {
                            if (it.isNotEmpty()) view.showProjects(it)
                            else view.showNoProjects()
                        },
                        // OnError
                        {
                            view.showConnectionError()
                        })
        subscriptions.add(subscription)
    }

    override fun openProjectDetails(project: Project) {
        view.showProjectDetails(project.id)
    }

    override fun subscribe() {}

    override fun unsubscribe() {
        subscriptions.clear()
    }
}
